pkg/apis/intevents/v1alpha1/brokercell_defaults.go

```diff
@@ -35,22 +63,67 @@ func (bc *BrokerCell) SetDefaults(ctx context.Context) {
+func (componentParams *ComponentParameters) SetMemoryDefaults(memoryLimitToRequestCoefficient float64, memoryRequest string) {
+	if componentParams.MemoryRequest == nil {
+		componentParams.MemoryRequest = ptr.String(memoryRequest)
+	}
+	requestedMemoryQuantity := resource.MustParse(*componentParams.MemoryRequest)
+	if componentParams.MemoryLimit == nil {
+		autoSelectedLimit := quantityutil.MultiplyQuantity(requestedMemoryQuantity, memoryLimitToRequestCoefficient)
+		componentParams.MemoryLimit = ptr.String(autoSelectedLimit.String())
+	}
+	// Make sure the limit is not lower than what's requested
+	memoryLimitQuantity := resource.MustParse(*componentParams.MemoryLimit)
+	if memoryLimitQuantity.Cmp(requestedMemoryQuantity) < 0 {
+		componentParams.MemoryLimit = ptr.String(requestedMemoryQuantity.String())
+		memoryLimitQuantity = requestedMemoryQuantity
+	}
+}
+
+// SetCPUDefaults sets the CPU consumption related default field values for ComponentParameters.
+func (componentParams *ComponentParameters) SetCPUDefaults(defaultCPURequest, defaultCPULimit string) {
+	if componentParams.CPURequest == nil {
+		componentParams.CPURequest = ptr.String(defaultCPURequest)
+	}
+	if componentParams.CPULimit == nil {
+		componentParams.CPULimit = ptr.String(defaultCPULimit)
 	}
-	if bcs.Components.Fanout.MaxReplicas == nil {
-		bcs.Components.Fanout.MaxReplicas = ptr.Int32(maxReplicas)
+}
+
+// SetAutoScalingDefaults sets the autoscaling-related default field values for ComponentParameters.
+func (componentParams *ComponentParameters) SetAutoScalingDefaults(targetMemoryUsageCoefficient float64, avgCPUUtilization int32) {
+	if componentParams.MemoryLimit == nil {
+		panic("Memory limit should be specified on the component before setting auto-scaling parameters")
 	}
-	if bcs.Components.Retry.MinReplicas == nil {
-		bcs.Components.Retry.MinReplicas = ptr.Int32(minReplicas)
+	if componentParams.MinReplicas == nil {
+		componentParams.MinReplicas = ptr.Int32(minReplicas)
 	}
-	if bcs.Components.Retry.MaxReplicas == nil {
-		bcs.Components.Retry.MaxReplicas = ptr.Int32(maxReplicas)
+	if componentParams.MaxReplicas == nil {
+		componentParams.MaxReplicas = ptr.Int32(maxReplicas)
 	}
-	if bcs.Components.Ingress.MinReplicas == nil {
-		bcs.Components.Ingress.MinReplicas = ptr.Int32(minReplicas)
+	if componentParams.AvgCPUUtilization == nil {
+		componentParams.AvgCPUUtilization = ptr.Int32(avgCPUUtilization)
 	}
-	if bcs.Components.Ingress.MaxReplicas == nil {
-		bcs.Components.Ingress.MaxReplicas = ptr.Int32(maxReplicas)
+	// If target average consumption for the auto-scaler is not explicitly specified or falls beyond reasonable range,
+	// default it based on the memory limit
+	memoryLimitQuantity := resource.MustParse(*componentParams.MemoryLimit)
+	if componentParams.AvgMemoryUsage == nil || memoryLimitQuantity.Cmp(resource.MustParse(*componentParams.AvgMemoryUsage)) < 0 {
```

Comment:
I think the second part belongs to validation.

Reply:
We can have a separate "validation" section, though to me it sounds more suitable for the cases when we won't proceed with given values and rather throw an exception immediately, whereas here it seemed fine to keep it with auto-selection/defaulting. Thoughts?

Reply:
See my other comment. We should rely on validation webhook. Auto correcting values provided by users would be confusing.